server/secure_test.go

```diff
@@ -0,0 +1,370 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/theplant/appkit/log"
+)
+
+//////////////////////////////////////////////////
+// CSRF Required Header Verification
+
+func ExampleCSRFHeader_Blank() {
+	cfg := CrossSiteConfig{
+		RawAllowedOrigins:  "http://example.com",
+		CSRFRequiredHeader: "", // Explicitly blank
+	}
+
+	s, req := setup(cfg)
+
+	req.Header.Add("origin", "http://example.com")
+
+	exec(s, req)
+
+	// Output: executed handler
+	// 200
+}
+
+func ExampleCSRFHeader_Set() {
+	cfg := CrossSiteConfig{
+		RawAllowedOrigins:  "http://example.com",
+		CSRFRequiredHeader: "X-Csrf", // Same as default
+	}
+
+	s, req := setup(cfg)
+
+	req.Header.Add("origin", "http://example.com")
+	req.Header.Add("X-Csrf", ".") // Value is ignored
+
+	exec(s, req)
+
+	// Output: executed handler
+	// 200
+}
+
+func ExampleCSRFHeader_Missing() {
+	cfg := CrossSiteConfig{
+		RawAllowedOrigins:  "http://example.com",
+		CSRFRequiredHeader: "X-Csrf", // Same as default
+	}
+
+	s, req := setup(cfg)
+
+	req.Header.Add("origin", "http://example.com")
+
+	exec(s, req)
+
+	// Output: level=warn msg="Request missing csrf header" during=appkit/server.verifyHeader
+	// 400
+}
+
+func ExampleCSRFHeader_CaseSensitivity() {
```

Comment:
I think the header is case **in**sensitivity. But the test title said the header is *case sensitivity*.

Reply:
It's not actually case insensitive, the header used in `CSRFRequiredHeader` must be capitalised. I'll add more to the example to demonstrate this.

Reply:
In the implemetation:

* the configuration value of `CSRFRequiredHeader` is case in-sensitivity, for example `x-csrf`, `X-csrf` and `X-CSRF` are all valid values.
* and in the request, the header also case in-sensitivity, for example `x-csrf`, `X-csrf` and `X-CSRF` are all valid header

But, in fact, the `verifyHeader` is using `X-Csrf` as *key* to get the request header, and the go's `net/http.Request` will parse the header using `http.CanonicalHeaderKey` too.

```sh
// The log that prints the `request.Header` in the source code
req.Header: map[User-Agent:[Go-http-client/1.1] Content-Length:[0] Origin:[http://example.com] X-Csrf:[.] Accept-Encoding:[gzip]]
```

Reply:
> But, in fact, the `verifyHeader` is using `X-Csrf` as *key* to get the request header, and the go's `net/http.Request` will parse the header using `http.CanonicalHeaderKey` too.

I think this is true only if you use [`net/http.Header.Get`](https://godoc.org/net/http#Header.Get):

> Get gets the first value associated with the given key. It is case insensitive; textproto.CanonicalMIMEHeaderKey is used to canonicalize the provided key.

And the middleware doesn't use `Get`. It treats header as `map[string][]string`, because it needs to know if the header was *present*, and the comment on `Get` says

> If there are no values associated with the key, Get returns ""

So we can't distinguish between *no* header and an *empty* header. I'm not sure if an empty header is valid HTTP though.

Reply:
> I think this is true only if you use net/http.Header.Get:

From the log, it looks the `request.Header`'s keys is canonicalized.

![image](https://user-images.githubusercontent.com/349931/40696693-727f1890-63f9-11e8-9a95-f4651d6a30b5.png)

Reply:
> From the log, it looks the request.Header's keys is canonicalized.

Sure, but if you use `Header` as a map, *your* key needs to be canonicalised too!

Reply:
> your key needs to be canonicalised too!

![image](https://user-images.githubusercontent.com/349931/40708110-343ccf1a-6425-11e8-9602-f93fb0b195ba.png)